source/common/network/io_socket_error_impl.h

```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include "envoy/api/io_error.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Network {
+
+class IoSocketError : public Api::IoError {
+public:
+  explicit IoSocketError(int sys_errno) : errno_(sys_errno) {}
+
+  ~IoSocketError() override {}
+
+  Api::IoError::IoErrorCode getErrorCode() const override;
+  std::string getErrorDetails() const override;
+
+private:
+  int errno_;
+};
+
+// IoErrorCode::Again is used frequently. Define it to be a singleton to avoid frequent memory
+// allocation of such instance. If this is used, IoHandleCallResult has to be instantiated with
+// deleter deleteIoError() below to avoid deallocating memory for this error.
+class IoSocketEagain : public IoSocketError {
+public:
+  IoSocketEagain() : IoSocketError(EAGAIN){};
+};
+
+inline IoSocketEagain* getIoSocketEagainInstance() {
+  static auto* kInstance = new IoSocketEagain();
+  return kInstance;
```

Comment:
This is probably a good place to put some kind of metrics or other data-driven indicating why this complexity is needed. It seems surprising that memory allocating a 16-byte struct is going to be significant compared with a system-call.

Reply:
I don't have any evidence that allocating IoError for EAGAIN can cost a lot of CPU here. But before abstracting out calls to fd, there is not heap allocation in read and write calls.Since IoError lives very shortly after a sys call, it's always a good practice to avoid such rapid alloc and dealloc.